### PR TITLE
typechecker: Choose broader type in arrays, sets and dicts

### LINKED ISSUE #1353: typechecker: Array type inference should infer base class instead of derived class when callee expects base class
This currently chokes in the C++ compiler:
```
class Base {}
class Derived: Base {}

fn func(anon a: [Base]) {}

fn main() {
    func([Derived()])
}
```
The problem is that `[Derived()]` creates a `[Derived]` when we really want `[Base]` since that is what the callee expects.

### PR DESCRIPTION
**Pull request by 0GreenClover0**
Fixes #1353

### CODE REVIEW

**Comment by 0GreenClover0** [comment]
Alternatively we can just always use the type hint id if it exists, I think

File: tests/typechecker/function_base_class_set_argument_derived_given.jakt [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/// Expect:
+/// - output: "PASS\n"
+
+trait Hashable<Output> {
+    fn hash(self) -> Output
+}
+
```
Comment (alimpfard) [inline tests/typechecker/function_base_class_set_argument_derived_given.jakt:7]: uhh...this definitely shouldn't be accepted, this is not the "correct" definition of Hashable as assumed by the compiler

Comment (alimpfard) [reply tests/typechecker/function_base_class_set_argument_derived_given.jakt:7]: I mean, defining that trait is fine, but `{T}` where `T implements(Hashable<i64>)` should not be accepted

Comment (0GreenClover0) [reply tests/typechecker/function_base_class_set_argument_derived_given.jakt:7]: Hm, so how this should look like?

Comment (alimpfard) [reply tests/typechecker/function_base_class_set_argument_derived_given.jakt:7]: `Hashable` is already in scope, imported from the prelude, it should just `implements(Hashable)`

Comment (0GreenClover0) [reply tests/typechecker/function_base_class_set_argument_derived_given.jakt:7]: It took me a sec to figure out I'm missing `this` in `public fn hash(this) => 42u32`
Now I know why it didn't work for me before :p

**Pull request merged**